src/node_main_instance.h

```diff
@@ -83,6 +83,7 @@ class NodeMainInstance {
   std::unique_ptr<ArrayBufferAllocator> array_buffer_allocator_;
   v8::Isolate* isolate_;
   MultiIsolatePlatform* platform_;
+  std::unique_ptr<v8::CppHeap> cpp_heap_;
```

Comment:
node_worker.h will likely need this also. Specifically, `v8::CppHeap` can only be used by a single thread at a time, and garbage collected objects managed by the heap are only safe to be cleaned up on the thread currently owning the lock for the heap. To be safe, we probably want each worker thread to have its own `v8::CppHeap` instance associated with its own `v8::Isolate`, and have some mechanism in place to ensure that oilpan gc'd objects are safely being handled cross-thread.

Reply:
Added to `node_worker.cc`: https://github.com/nodejs/node/pull/45704/files#diff-b06b65a7f7e33816a1986d731765d732ab23f585bece6b23c5e83ca3682dba62

Regarding
> and have some mechanism in place to ensure that oilpan gc'd objects are safely being handled cross-thread

If the worker threads have an entirely separate `v8::Isolate`, then there's no cross-thread object sharing that can take place, and we don't need to do anything?

I haven't used worker threads, so I could be wrong, but taking a brief scan of the docs it looks like the object passing is through `ArrayBuffer`, either by transferring ownership of the underlying native memory (in which case the local object becomes empty), or by cloning the underlying native memory (if marked untransferable).  In neither case are the v8 objects being transferred/shared, just the backing data?

Reply:
We have cases where we share underlying objects across multiple threads. For instance, cloned key object data, or Blob internal storage, etc. We basically just need to make sure those type things aren't GC managed, and keep the GC managed things local to their owning threads.

Reply:
For cppgc objects the objects are managed by V8, I don't think we need to do anything else. We only need to make sure that the `CppHeap` is alive while the isolate is alive, and that's why we'd need to own the pointer. But the reference to the `CppHeap` should be retrieved via `isolate->GetCppHeap()` instead, and before that the user should've already ensured that they are accessing `isolate` on the right thread.

> We basically just need to make sure those type things aren't GC managed, and keep the GC managed things local to their owning threads.

We don't have any internal objects that are GC managed for now. That would need to be explicitly implemented.